docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,298 @@
+
+### Permissions
+
+The client does not directly/automatically handle permissions needed for the arbitrary host it could run on.
+There is no expectation that the deployer of the host is also using wing so there is no shared system to transfer permission knowledge.
+
+For example, when using
+
+```wing
+client.lift(lift { bucket: [put] });
+```
+
+it would be helpful to retrieve the necessary information about the `bucket` and the `put` operation.
+
+If you're using terraform and deploying to AWS, you will need the bucket name and the correspond IAM policy.
+There is no abstract API available to store/retrieve this information currently as this is handled only by target-specific subtypes of `IInflightHost`.
+
+To keep this simple, we can provide and a helper API in the shared aws module (and other platforms, as needed) to generate the necessary policies by exposing the lift information (i.e. the `_liftMap`) from the client.
+
+```wing
+bring aws;
+
+let policies = aws.generatePolicyStatements(client.lifts);
+```
+
+With these, one can generate a policy document to attach to a role/user to be used by the client's host.
+
+If you're using the simulator, this problem is (currently) more difficult.
+The simulator is centralized and will have no knowledge of the future clients that will be using it.
+The simulator has no API to add new permissions after the simulator has started and the scope of that feels too large for this RFC.
+
+As an MVP, the simulator can be configured to ignore permissions via `wing.toml`.
+Even though this is unsafe and cause cause local/cloud differences, wing is intentionally not claiming ownership of the client's host environment.
+
+### Type generation with `lift`
+
+Wing's compiler has type information available for use, but only before preflight execution occurs.
+This makes an API like the following not feasible to implement using logic similar to the existing `extern` or `@inflight`:
+
+```wing
+client.lift("bucket", bucket, ["put"]);
+```
+
+While executing preflight, the knowledge about the type of the `bucket` instance is lost, so generating the actual client code will have no way to reference it.
+
+This RFC proposes changing the existing `lift` syntax as a way to capture the necessary type information during preflight.
+An important feature of this syntax is that it's entirely statically analyzable, so anything we need to do with it can be done before-preflight.
+The existing `lift` syntax is currently intended to be used only inside inflights to explicitly lift preflight data into the defined scope:
+
+```wing
+let b1 = new cloud.Bucket();
+inflight () => {
+  let var b = b1;
+  lift { b: [put] } {
+    b.put("key", "value");
+  }
+}
+```
+
+With some changes, with syntax could also be used as a way to create a preflight container for lifted data.
+
+```wing
+let b = new cloud.Bucket();
+let lifts = lift { b: [put] };
```

Comment:
I considered proposing `@lift({ b: [put] })`, but I'm not sure it was valuable to discuss this change here since really effect the behavior. I expect the change about what `lift` can do to be more controversial anyways.